Show progress bar in buttons when clicked

Protoss78 opened the pull request:
This PR should fix issue #64 . Instead of showing a spinner I decided to use a progress bar at the bottom of each button - mainly because I felt it is simpler to style from a layout perspective. The spinner was constantly changing the button size when it appeared.
The progress bar should appear on the following buttons:

- All social logins
- Email sign in
- Email sign up
- Password reset

I added a new `_inProgress` property to _login-fire-common-behavior_ to realize that. Whenever a sign in related process is started, the property is set to true. When a process completes successfully or fails, this property is set back to false. 
The progress bar and its associated buttons then simply react to the state change of this property. 
Styling samples were added to _login-fire_ and _login-fire-social_ pages.

Protoss78 committed:
Show progress-bar while signing in or out. Initial version to fix issue #64

Protoss78 committed:
Refactored code and prevented additionl click events when already in progress

Protoss78 committed:
feat: Added progress bar to buttons on email login form

Protoss78 committed:
feat: Added progress bar styling samples to demo

masonlouchart commented:
Thanks for your contribution! We'll triage and take a look at it as soon as possible!

masonlouchart commented on demo/login-fire-social.html:
```diff
@@ -131,13 +131,15 @@ <h3>Max horizontal, no icons login-fire-social Demo</h3>
         login-fire-social.custom {
           width: 100%;
+          --paper-progress-active-color: darkred;
+          --paper-progress-secondary-color: lightred;
```
The demo does not use the CSS custom variable `--login-fire-button-progress` introduced with this commit. Was it intended?

Protoss78 replied:
Yes this is intentional. I wanted to treat it the same way as styling paper-input and use the existing paper-progress variables. By mistake I forgot to remove the --login-fire-button-progress variable I had at one point in time. Will remove that instead.

Protoss78 replied:
Ok. While looking into the whole styling I added the rounded corners to the progress bar and added default styling for all social login buttons. But if we want to override those values, I might have to introduce CSS custom variables again. Will look into that. :smirk:

masonlouchart commented on login-fire-button.html:
```diff
@@ -130,15 +131,24 @@
         @apply --login-fire-btn-signout-hover;
       }
 
+      .btn-progress {
```
This CSS class can be moved into the `login-fire-styles` file and the shared styles included in the `login-fire-form` element. But is it really necessary to import the shared styles in the form just to have the progress bar one?... 🤔

Protoss78 replied:
I will move it to the shared styles files, as it is better to maintain the default styling at a central place instead of copying it.

masonlouchart commented on login-fire-common-behavior.html:
```diff
@@ -291,6 +291,17 @@
+      },
+
+      /**
+       * The _inProgress indicator shows whether a sign in or sign out process is currently running.
```
As recommended in the [Polymer documentation](https://www.polymer-project.org/2.0/docs/tools/documentation#language), try to follow the [JavaDoc style](http://www.oracle.com/technetwork/java/javase/documentation/index-137868.html#styleguide). I would have wrote something like: _"Indicates if a sign in or sign out process is in progress."_

Protoss78 replied:
Agree. Will change it.

masonlouchart commented on login-fire-common-behavior.html:
```diff
@@ -319,6 +330,8 @@
       }
       if (!this.noSignOut && this.signedIn) {
+        // Set _inProgress flag to true when a signout is started
+        this._inProgress = true;
```
Set back the `_inProgress` property to `false` is not required here. It will be done inside the `_fireEvent` callback where it's a better place to do so.

Protoss78 replied:
I'm confused about this comment? Is it maybe misplaced? _inProgress is set to true here. It will be set back once the success or error event is fired.

masonlouchart replied:
What I am talking about here... 😒 . I check this part again.

masonlouchart replied:
Please @Protoss78 , forgive me. It was 4AM in France when I wrote this shitty comment. 💩

Protoss78 replied:
No problem at all. :+1:

masonlouchart commented on login-fire-form.html:
```diff
@@ -441,12 +455,14 @@
           }.bind(this), function(error) {
             this.set('_lastError', error.message);
+            this._inProgress = false;
           }.bind(this));
```
Maybe it's possible to replace the "new" lines 462 and 465 by a call to the [Promise.prototype.finally](https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Objets_globaux/Promise/finally) method on the response of the `sendPasswordResetEmail()`?

Something like:
```javascript
this.$$('#auth').auth.sendPasswordResetEmail(this.email)
	.then(function(resp){/*...*/}.bind(this), function(error){/*...*/}.bind(this))
	.finally(function(){/*...*/ this._inProgress = false;}.bind(this));
```

Protoss78 replied:
Yep. Still new to Promise land. :grin: Will change that one as well.

masonlouchart replied:
I took a look at your repos and I can guess what you're made of 😉 !

🆒 Keep Calm & Contribute 😎

masonlouchart commented on login-fire-social-behavior.html:
```diff
@@ -78,6 +80,13 @@
      */
     signInOrOut: function() {
+      // Skip processing of this method when we already have started a sign or signout process
+      if (this._inProgress) {
```
This check must be done at the lowest level possible to avoid redo it several times. The `signIn` and `signOut` methods have to check the progress flag, not the callers of those methods.

Protoss78 replied:
Agree. Will change it and move it to the low level methods.

masonlouchart reviewed (changes requested):
Some minor changes to do before to approve the PR. Please see the comments.

### Additional work to be perfect (optional):
- the progress bar could have the same color of the background of the button
- the bottom corners of the button are not rounded anymore when the progress bar is displayed

Protoss78 committed:
Added default styling for social logins and moved progress bar style to shared styles

Protoss78 committed:
Updated documentation, moved double click check to low level methods and added use of finally.

Protoss78 committed:
Customized styling of paper-progress colors, added default values and added the new styles to documentation.

Protoss78 commented:
Ok. Looks like I'm done with my round of updates. Thanks for the review and the feedback. :+1:

masonlouchart commented:
Thank you for your help to improve this element. I'll take a look at the updates soon before merge.

masonlouchart commented:
CSS custom variables for the progress bar are not applied with Polymer 1 😕 . Because there is still some issues with styling we can merge this PR.

masonlouchart committed:
chore(log): reorganize logs in signOut method

The first log of the method signOut was move after the check of the
"inProgress" flag to minimize useless check. Also, the "console.log" was
turned into a "console.debug" to keep consistency in the source code.
